course: task 2.3 form validator

BorislavAnchev opened the pull request:
The form validator task.

mgenov commented on 23FormValidator/validation.js:
```diff
@@ -0,0 +1,69 @@
+    && /.{1,100}/.test(this.address) // Try without ^ and $.
+    && /^.{6,18}$/.test(this.password)
+    && ( this.password === this.confirmPassword)) {
+      alert('It works.')
```
It would be better fields to be highlighted or something instead of showing alerts to the user. 

Note that Form is a common name but you are verifying concrete fields which doesn't sound good.

BorislavAnchev replied:
Is it bad to use the 'pattern' attribute?

mgenov replied:
It's not bad, but the goal of the task is to improve your JS skills.

BorislavAnchev replied:
Fixes are available for review.

mgenov commented on 23FormValidator/validation.js:
```diff
@@ -0,0 +1,69 @@
+inputFields.forEach(function(item) {
+  document.getElementById(item).addEventListener('change', function() {
+    switch(this.id) {
+      case 'firstName': {
```
Looks like duplication with input fields. Better design would be each field validation to be separate instance. 

```jsx
const registrationFields = [
   new Field('firstName', new RegExValidator(/^[A-Z]{1}[a-z]{0,14}$/), 'Name must start with upper case')
]
```

BorislavAnchev commented:
New changes are ready for review:

- Created a new 'RegExValidator' class, based on the given example
- Made corrections to the former 'Form' class:
  - The class is now named 'Field' in order to better distinguish it from the <form> tag and other common elements
  - Changed the class' internal fields and constructor parameters to correspond to the given example. Now the constructor receives new instance of RegExValidator as second argument on instantiation
  - Tweaked the 'validate()' method of the class in order be able to work with different instances of the class and different contexts
  - The event listeners now work with individual class instances for each input field
- Improved the RegEx for the address field
- Added alert fields for each of the input fields
- Removed the button at the bottom of the form as it now serves no function since the form is not submitted anywhere at this moment, and every input field triggers it's validation on change rather than on button click like how it was before

BorislavAnchev commented:
@mgenov Code ready for review.

mgenov commented on 23FormValidator/validation.js:
```diff
@@ -0,0 +1,55 @@
+  }
+
+  validate(text) {
+    if(this.id === 'age') {
```
This check shouldn't belong here as it's concrete but the class is common. If you have such check you have to name your field AgeField ...

mgenov commented on 23FormValidator/validation.js:
```diff
@@ -0,0 +1,55 @@
+  }
+}
+
+class Field {
```
As field class accepts validation regex, the validation should use only this regex.

BorislavAnchev commented:
Changes are ready for a review.

mgenov commented on 23FormValidator/validation.js:
```diff
@@ -0,0 +1,57 @@
+];
+
+registrationFields.forEach(function(field) {
+  document.getElementById(field.id).addEventListener('change', function() {
```
Doesn't look correct to be like that. The goal of the field is to provide validation on its own.

BorislavAnchev commented:
Changes ready for a review.
